src/util.cc

```diff
@@ -1,8 +1,196 @@
 #include "util.h"
 #include "string_bytes.h"
 
+#define UNI_SUR_HIGH_START    (uint32_t) 0xD800
+#define UNI_SUR_LOW_END       (uint32_t) 0xDFFF
+#define UNI_REPLACEMENT_CHAR  (uint32_t) 0x0000FFFD
+#define UNI_MAX_LEGAL_UTF32   (uint32_t) 0x0010FFFF
+
+#if __GNUC__ >= 4 || (__GNUC__ == 3 && __GNUC_MINOR__ >= 4)
+#define HAS_GCC_BUILTIN_CLZ
+#endif
+
+#ifdef WIN32
+#include <intrin.h>
+static uint32_t __inline clz(uint32_t xs) {
+  unsigned long result = 0;
+  _BitScanReverse(&result, xs);
+  return (31 - result);
+}
+#elif defined(HAS_GCC_BUILTIN_CLZ)
+#define clz(xs) __builtin_clz(xs)
+#else
+static inline uint32_t clz(uint32_t xs) {
+  uint32_t out = 0;
+  while (xs >> (31 - out)) {
+    ++out;
+  }
+  return out;
+}
+#endif
+
 namespace node {
 
+typedef size_t (*ErrorStrategy)(
+    const size_t, const char*, const size_t);
+typedef void (*GlyphStrategy)(
+    const char*, const size_t, const uint32_t, const size_t);
+
+
+static size_t Skip(
+    const size_t remaining, const char* input, const size_t glyph_size) {
+  if (remaining > glyph_size) {
+    return 1;
+  }
+  return 0;
+}
+
+
+static size_t Halt(
+    const size_t remaining, const char* input, const size_t glyph_size) {
+  return 0;
+}
+
+
+static void DiscardGlyph(
+    const char* glyph,
+    size_t glyph_size,
+    uint32_t glyph_value,
+    const size_t pos) {
+}
+
+
+static bool IsLegalUTF8Glyph(const uint8_t* input, const size_t length) {
+  uint8_t acc;
+  const uint8_t* srcptr = input + length;
+  switch (length) {
+    default: return false;
+    case 4:
+      acc = (*--srcptr);
+      if (acc < 0x80 || acc > 0xBF)
+        return false;
```

Comment:
Can you add `// Fall through.` comments?  That makes it clear to the reader that the lack of `break` is intentional.